Radio.m

```diff
@@ -1445,14 +1591,66 @@ - (void) parseGpsToken: (NSScanner *) scan {
     
 }
 
-
+//
+// Parse Profile tokens
+//     called on the GCD thread associated with the GCD tcpSocketQueue
+//
+//     format: <apiHandle>|profile <profileType> list=<value>^<value>^...<value>^
+//                           OR
+//     format: <apiHandle>|profile <profileType> current=<value>
+//
+//     scan is initially at scanLocation = 17, start of the <profileType>
+//     "<apiHandle>|profile " has already been processed
+//
 - (void) parseProfileToken: (NSScanner *) scan {
-    
+    // get the Profile type
+    NSString *profileType;
+    [scan scanUpToString:@" " intoString: &profileType];
+    // skip the space
+    [scan scanString:@" " intoString: nil];
+    
+    // get the Sub type
+    NSString *profileSubType;
+    [scan scanUpToString:@"=" intoString: &profileSubType];
+    // skip the "="
+    [scan scanString:@"=" intoString: nil];
+    
+    // get the remainder of the command
+    NSString *remainderOfCommand;
+    [scan scanUpToString:@"\n" intoString: &remainderOfCommand];
+    
+    // List or Current value?
+    if ([profileSubType isEqualToString: @"list"]) {
+        // it's the List, separate the components
+        NSMutableArray *profileNames;
+        profileNames = [[remainderOfCommand componentsSeparatedByString:@"^"] mutableCopy];
+        // remove the last (empty) string
+        [profileNames removeLastObject];
+        // save it in the appropriate property
+        if ([profileType isEqualToString: @"global"]) {
+            @synchronized (self.globalProfiles) {
+                _globalProfiles = profileNames;
```

Comment:
> On Jul 8, 2015, at 7:18 PM, Stu Phillips <notifications@github.com> wrote:
> 
> Also, I notice that flex uses the caret ^ to replace spaces in profile names. That should be handled transparently to our client here. Both for setting the names in the arrays plus for removing spaces when the profile name is sent to the radio

The caret isn’t replacing spaces, it’s used as the delimiter between Profile Names (which may contain spaces, e.g. “Default FHM-1 DX”). A typical profile status string looks like this:

	S78613D02|profile tx list=Default^Default FHM-1^Default FHM-1 DX^Default HM-Pro^Default PR781^. . . ^RadioSport WIDE M360/EM56^


I can see why they did this, it allows you to have Profile Names with embedded spaces (but no carets). As long as we send the messages back to the Radio with caret delimiters we should be good.

A similar approach was taken in the versionResponseCallback message where they used hash (“#”) as the delimiter between “key=value” pairs. A typical one of those looks like this:

	R13|0|SmartSDR-MB=1.4.11.60#PSoC-MBTRX=1.3.1.0#PSoC-MBPA100=1.1.2.2#FPGA-MB=0.0.23.167

I’m not sure why they did it here but the good news is we never send a message like this to the Radio.

73’s 
Doug - K3TZR